src/storage.rs

```diff
@@ -0,0 +1,135 @@
+        Address(self.0 + other.0)
+    }
+}
+
+/// Page represents an unsigned int that is a Page ID in the device memory space.
+pub struct Page<U>(U);
+
+/// Read a single word from the device.
+///
+/// `Word` type allows any word size to be used.
+pub trait SingleRead<Word, U> {
+    /// An enumeration of Storage errors
+    type Error;
+
+    /// Reads the word stored at the address
+    /// ```
+    /// pub fn try_read(&mut self, address: Address) -> nb::Result<u8, Self::Error>
+    ///     let address = address.0 as *const _;
+    ///     unsafe {
+    ///         Ok(core::slice::from_raw_parts::<'static, u8>(address,length))
+    ///     }
+    /// ```
+    fn try_read(&mut self, address: Address<U>) -> nb::Result<Word, Self::Error>;
+}
+
+/// Write a single word to the device.
+///
+/// `Word` type allows any word size to be used.
+pub trait SingleWrite<Word, U> {
+    /// An enumeration of Storage errors
+    type Error;
+
+    /// Writes the word to the address
+    fn try_write(&mut self, address: Address<U>, word: Word) -> nb::Result<(), Self::Error>;
+}
+
+/// Read multiple bytes from the device.
+///
+/// Intended to be used for when there is a optimized method of reading multiple bytes.
+///
+/// Iterating over the slice is a valid method to ```impl``` this trait.
+pub trait MultiRead<Word, U> {
+    /// An enumeration of Storage errors
+    type Error;
+
+    /// Reads the words stored at the address to fill the buffer
+    /// ```
+    /// pub fn try_read_slice(
+    ///     &mut self,
+    ///     address: Address,  
+    ///     buf: &mut [Word]
+    /// ) -> nb::Result<Word, Self::Error>
+    ///     let address = address.0 as *const _;
+    ///     unsafe {
+    ///         buf = core::slice::from_raw_parts::<'static, Word>(address,buf.len())
+    ///     }
+    ///     
+    ///     Ok()
+    /// }
+    /// ```
+    fn try_read_slice(
+        &mut self,
+        address: Address<U>,
+        buf: &mut [Word],
+    ) -> nb::Result<(), Self::Error>;
+}
+
+/// Write multiple bytes to the device.
+///
+/// Intended to be used for when there is a optimized method of reading multiple bytes.
+///
+/// Iterating over the slice is a valid method to ```impl``` this trait.
+pub trait MultiWrite<Word, U> {
+    /// An enumeration of Storage errors
+    type Error;
+
+    /// Writes the buffer to the address
+    fn try_write_slice(&mut self, address: Address<U>, buf: &[Word])
+        -> nb::Result<(), Self::Error>;
+}
+
+/// A common interface to erase pages or memory locations.
+///
+/// For Flash storage, the write functions can't set a bit to 1.
+///
+/// For non flash devices, this trait is not required, but it can be used to erase data as recommended by the device (EG set all fields to 0).
+pub trait ErasePage<U> {
+    /// An enumeration of Storage errors
+    type Error;
+
+    /// Erase the page of memory
+    /// For flash devices, this sets the whole page to 0xFF
```

Comment:
I also think the reset value is probably better left to implementations. Anything written in the documentation of the trait would be part of a contract. (also below)